File: pythonlibrary/aibrd/generators/ambiguity_report.py
from datetime import date
from ..llm.client import call_llm_json
from ..models.outputs import Ambiguity

_SYSTEM = """You are a business analyst reviewing a BRD for ambiguities.
Flag vague terms lacking measurable definitions (e.g. "fast", "intuitive", "soon", "large").
Return JSON: {"ambiguities": [{"term": "...", "context": "...", "suggestion": "..."}]}"""


def detect_ambiguities(brd_text: str) -> list[Ambiguity]:
    raw = call_llm_json(brd_text, _SYSTEM)
    return [
        Ambiguity(id=f"AMB-{i+1:03d}", **a)
        for i, a in enumerate(raw.get("ambiguities", []))
    ]


def format_ambiguity_report(ambiguities: list[Ambiguity]) -> str:
    today = date.today().isoformat()
    lines = [
        "# Ambiguity Report",
        f"_Generated: {today}_",
        "",
        "> Terms flagged as vague or unmeasurable. Resolve before sprint planning.",
        "",
    ]
    if not ambiguities:
        lines.append("No ambiguities detected.")
        return "\n".join(lines)

    for a in ambiguities:
        lines += [
            f'## {a.id}: "{a.term}"',
            f"**Found in:** _{a.context}_",
            f"**Suggestion:** {a.suggestion}",
            "",
        ]
    return "\n".join(lines)

File: vscodebase/src/core/generators/ambiguity_report.ts
import * as vscode from 'vscode'
import { callLLMJson } from '../../llm/client'
import { Ambiguity } from '../models/outputs'

const SYSTEM = `You are a business analyst reviewing a BRD for ambiguities.
Flag vague terms that lack measurable definitions (e.g. "fast", "intuitive", "soon", "large", "efficient").
Return JSON: {
  "ambiguities": [{
    "term": "the vague term",
    "context": "the sentence it appears in",
    "suggestion": "how to make it measurable"
  }]
}`

export async function detectAmbiguities(
  brdText: string,
  token?: vscode.CancellationToken
): Promise<Omit<Ambiguity, 'id'>[]> {
  const result = await callLLMJson<{ ambiguities: Omit<Ambiguity, 'id'>[] }>(
    brdText,
    SYSTEM,
    token
  )
  return result.ambiguities
}

export function formatAmbiguityReport(ambiguities: Ambiguity[]): string {
  const lines: string[] = [
    '# Ambiguity Report',
    `_Generated: ${new Date().toISOString().split('T')[0]}_`,
    '',
    '> Terms flagged as vague or unmeasurable. Resolve before sprint planning.',
    ''
  ]

  if (ambiguities.length === 0) {
    lines.push('No ambiguities detected.')
    return lines.join('\n')
  }

  for (const a of ambiguities) {
    lines.push(`## ${a.id}: "${a.term}"`)
    lines.push(`**Found in:** _${a.context}_`)
    lines.push(`**Suggestion:** ${a.suggestion}`)
    lines.push('')
  }

  return lines.join('\n')
}
